src/meta-srv/src/selector/weight_based.rs

```diff
@@ -0,0 +1,200 @@
+
+        // 2. get stat kvs and filter out expired datanodes.
+        let stat_keys = lease_kvs.keys().map(|k| k.into()).collect();
+        let stat_kvs = filter_out_expired_datanode(
+            ctx.meta_peer_client.get_dn_stat_kvs(stat_keys).await?,
+            &lease_kvs,
+        );
+
+        // 3. make sure the regions of a table are distributed on different datanodes.
+        let stat_kvs = if let (Some(catalog), Some(schema), Some(table)) =
+            (&ctx.table, &ctx.schema, &ctx.table)
+        {
+            filter_out_datanode_by_table(stat_kvs, catalog, schema, table)
+        } else {
+            stat_kvs
+        };
+
+        // 4. compute weight array.
+        let weight_array = self.weight_compute.compute(&stat_kvs);
+
+        // 5. choose peers by weight_array.
+        let mut weighted_choose = self.weighted_choose.write();
+        let selected = choose_peers(weight_array, &opts, &mut *weighted_choose)?;
+
+        Ok(selected)
+    }
+}
+
+pub fn filter_out_expired_datanode(
+    mut stat_kvs: HashMap<StatKey, StatValue>,
+    lease_kvs: &HashMap<LeaseKey, LeaseValue>,
+) -> HashMap<StatKey, StatValue> {
+    lease_kvs
+        .iter()
+        .filter_map(|(lease_k, _)| stat_kvs.remove_entry(&lease_k.into()))
+        .collect()
+}
+
+pub fn filter_out_datanode_by_table(
+    stat_kvs: HashMap<StatKey, StatValue>,
+    catalog: &str,
+    schema: &str,
+    table: &str,
+) -> HashMap<StatKey, StatValue> {
+    stat_kvs
+        .into_iter()
+        .filter(|(_, stat_v)| contains_table(stat_v, catalog, schema, table) != Some(true))
+        .collect()
+}
+
+pub fn choose_peers<W>(
+    weight_array: Vec<WeightedItem<Peer>>,
+    opts: &SelectorOptions,
+    weighted_choose: &mut W,
+) -> Result<Vec<Peer>>
+where
+    W: WeightedChoose<Peer>,
+{
+    let min_required_items = opts.min_required_items;
+    ensure!(
+        !weight_array.is_empty(),
+        error::NoEnoughAvailableDatanodeSnafu {
+            required: min_required_items,
+            available: 0_usize,
+        }
+    );
+
+    let weight_array_len = weight_array.len();
+    weighted_choose.set_weight_array(weight_array)?;
+
+    if opts.allow_duplication {
+        (0..min_required_items)
+            .map(|_| weighted_choose.choose_one())
+            .collect::<Result<_>>()
+    } else {
+        // When opts.allow_duplication is false, we need to check that the length of the weighted array is greater than
+        // or equal to min_required_items, otherwise it may cause an infinite loop.
+        ensure!(
+            weight_array_len >= min_required_items,
+            error::NoEnoughAvailableDatanodeSnafu {
+                required: min_required_items,
+                available: weight_array_len,
+            }
+        );
+
+        let mut selected = HashSet::with_capacity(min_required_items);
+        while selected.len() < min_required_items {
```

Comment:
Will this loop run for a long time?

Reply:
We can do it in the following way:

Assume min_required_items is 3, weight_array_len is 100, then we can select 3 items from the weight array and return. 

But assume min_required_items is 99, weight_array_len is 100. It's not cheap to choose 99 items from the weight array. So we can choose 1 item from the weight array, and return the remaining 99 items.